### PR TITLE
Added paymaster feature

### LINKED ISSUE #700: [4337] Allow paymaster to cover fees
## Context / issue

One of the benefits of using 4337 is the possibility to use a paymaster to pay the fees on behalf of the account, instead having the native token available in the smart account.

This can enable different scenarios, like being able to create transactions without having funds in an account or being able to use ERC20 tokens to pay the gas fees

## Proposed solution

Create the user operation (safe operation) setting a paymaster and test that is possible to do transactions without funds, or paying using ERC20 tokens.

### PR DESCRIPTION
**Pull request by DaniSomoza**
## What it solves
Resolves #700 

### Counterfactual deployment tests

1. Counterfactual Deployment using `xDAI` held by the Safe Account in Gnosis chain .
2. Counterfactual Deployment using ERC20 (`USDC`) held by the Safe Account in Gnosis chain (see [Pimlico docs](https://docs.pimlico.io/paymaster/erc20-paymaster/faqs#step-3-generate-the-paymasteranddata-to-place-in-your-useroperation)).
3. Counterfactual Deployment using an sponsored transaction in Gnosis Chain.
4. Counterfactual Deployment using an sponsored transaction and a policy in Gnosis Chain.

##  Counterfactual Deployment using `xDAI` (Held by the Safe Account)

``` typescript

  // 1) Initialize pack with the paymaster data
  const safe4337Pack = await Safe4337Pack.init({
    ethersAdapter,
    rpcUrl: RPC_URL,
    bundlerUrl: BUNDLER_URL,
    options: {
      owners: [await signer.getAddress()],
      threshold: 1,
      saltNonce: '4337'
    }
  })
```

### First batch (Safe deployment + 2 USDC transfers)

**Deployment of a 4337 Safe Account using the xDAI held by the Safe in Gnosis chain and executing a batch of 2 transactions (2 USDC transfers)**

script used: `usdc-transfer-4337-counterfactual.ts`

NOTE: To execute this script, I needed to send at least `0.1 xDAI` to the Safe's address.

Safe Deployed: [`0x5370Dc00f52ef83BB7F126Ad02C129dE48a21A26`](https://app.safe.global/transactions/history?safe=gno:0x5370Dc00f52ef83BB7F126Ad02C129dE48a21A26)

Transaction:  [`0xf40225be749ac794f6ad283f533046be70a6615bad65c8c7216be47a0bde7d00`](https://gnosisscan.io/tx/0xf40225be749ac794f6ad283f533046be70a6615bad65c8c7216be47a0bde7d00)

JiffyScan: https://jiffyscan.xyz/userOpHash/0x046de6fe121ba0ba19cee927aede721c0a470ef60f34fa999a858f97c313769b?network=gnosis

### Second batch (2 USDC transfers with the already deployed Safe Account)

After that I executed again the script `usdc-transfer-4337-counterfactual.ts`

Transaction:  [`0x27efa8209471170fbed3a5685b5b283d229c2b081e5b527efa6e2507edea19d3`](https://gnosisscan.io/tx/0x27efa8209471170fbed3a5685b5b283d229c2b081e5b527efa6e2507edea19d3)

JiffyScan: https://jiffyscan.xyz/userOpHash/0x8cb50c681643d1a00df0482d370c527416ed867b3b4e58c7933b2c7a088148d2?network=gnosis

##  Counterfactual Deployment using ERC20 tokens (`USDC`) (Held by the Safe Account)

``` typescript

  // 1) Initialize pack with the paymaster data
  const safe4337Pack = await Safe4337Pack.init({
    ethersAdapter,
    rpcUrl: RPC_URL,
    bundlerUrl: BUNDLER_URL,
    paymasterOptions: {
      paymasterTokenAddress: usdcTokenAddress,
      paymasterAddress
      // amountToApprove?: bigint // optional value to set the paymaster approve amount on the deployment
    },
    options: {
      owners: [await signer.getAddress()],
      threshold: 1,
      saltNonce: '4337'  // to update the address
    }
  })
```
script used: `usdc-transfer-4337-erc20-counterfactual.ts`

### First batch (Safe deployment + 2 USDC transfers)

**Deployment of a 4337 Safe Account using the ERC20 (`USDC`) held by the Safe in Gnosis chain and executing a batch of 2 transactions (2 USDC transfers)**

Safe Deployed: [`0x4b2B32129992891d0d11dA4C2aC90ca462DbCe22`](https://app.safe.global/transactions/history?safe=gno:0x4b2B32129992891d0d11dA4C2aC90ca462DbCe22)

Transaction: [`0xb6223fbcbac6b1d66c90e5aebc00ab21c2952e2f7fc69fbd318cbdb181869150`](https://gnosisscan.io/tx/0xb6223fbcbac6b1d66c90e5aebc00ab21c2952e2f7fc69fbd318cbdb181869150)

JiffyScan: https://jiffyscan.xyz/userOpHash/0xdaf7c7b6895ad256bfe805e2f774666f22e48e781be0fa1ada8b47c7ae62cf42?network=gnosis

### Second batch (2 USDC transfers with the already deployed Safe Account)

After that I executed again the script `usdc-transfer-4337-erc20-counterfactual.ts`

Transaction:  [`0x27efa8209471170fbed3a5685b5b283d229c2b081e5b527efa6e2507edea19d3`](https://gnosisscan.io/tx/0x27efa8209471170fbed3a5685b5b283d229c2b081e5b527efa6e2507edea19d3)

JiffyScan: https://jiffyscan.xyz/userOpHash/0x8cb50c681643d1a00df0482d370c527416ed867b3b4e58c7933b2c7a088148d2?network=gnosis

## Counterfactual Deployment using an sponsored transaction in Gnosis Chain.

``` typescript

  // 1) Initialize pack with the paymaster data
  const safe4337Pack = await Safe4337Pack.init({
    ethersAdapter,
    rpcUrl: RPC_URL,
    bundlerUrl: BUNDLER_URL,
    paymasterOptions: {
      isSponsored: true,
      paymasterUrl: PAYMASTER_URL,
      paymasterAddress,
      sponsorshipPolicyId: POLICY_ID // this is an optional parameter
    },
    options: {
      owners: [await signer.getAddress()],
      threshold: 1,
      saltNonce: '4337'
    }
  })
```
script used: `usdc-transfer-4337-sponsored-counterfactual.ts`

### First batch (Safe deployment + 2 USDC transfers)

**Deployment of a 4337 Safe Account using my current balance in [the Pimlico payments Dashboard](https://dashboard.pimlico.io/payment) and executing a batch of 2 transactions (2 USDC transfers)**

Safe Deployed: [`0x57a824C1375078F6e1427B3CB1d30c8822167800`](https://app.safe.global/transactions/history?safe=gno:0x57a824C1375078F6e1427B3CB1d30c8822167800)

Transaction: [`0x75fe4ab7d65d35a277e9500e4f07e694a5566b12f236cd8d12e0c751e37ce8a5`](https://gnosisscan.io/tx/0x75fe4ab7d65d35a277e9500e4f07e694a5566b12f236cd8d12e0c751e37ce8a5)

JiffyScan: https://jiffyscan.xyz/userOpHash/0xa920a2bc8176b0af3cd76459f8bcf322996935e9180b72c3eea64176a68e1928?network=gnosis

My balance in Pimlico has decreased by an amount equivalent to the transaction cost on Gnosis chain:
Captura de pantalla 2024-03-14 a las 17 15 03

### Second batch (2 USDC transfers with the already deployed Safe Account)

After that I executed again the script `usdc-transfer-4337-sponsored-counterfactual.ts`

Transaction:  [`0x05b3298f7b1a983e9be8636dc0d7472a7a133317ff7850c170a9b869a9d02583`](https://gnosisscan.io/tx/0x05b3298f7b1a983e9be8636dc0d7472a7a133317ff7850c170a9b869a9d02583)

JiffyScan: https://jiffyscan.xyz/userOpHash/0x20270254c16238c458c079630da1694978cf280971f7ce666c0b2ea32c9514b2?network=gnosis

## Counterfactual Deployment using an sponsored transaction and a policy in Gnosis Chain.

Policy Id used: `sp_pink_marten_broadcloak`
Policy: Only one transaction per user.

First transaction: [`0x37980b61cfec64d5aa1182d8506463f77082192c2fa448756d74d9305140f19b`](https://gnosisscan.io/tx/0x37980b61cfec64d5aa1182d8506463f77082192c2fa448756d74d9305140f19b)

Second transaction: `sponsorshipPolicy perUserMaximumOpCount reached` error:

Captura de pantalla 2024-03-15 a las 10 11 59

## How this PR fixes it

Added paymasterOptions: 

- `isSponsored`: flag to use sponsored transactions with Pimlico. (see your current balance in [the Pimlico payments Dashboard](https://dashboard.pimlico.io/payment))
- `sponsorshipPolicyId`: optional parameter to specify your policy. You can create your policy in [the Pimlico dashboard](https://dashboard.pimlico.io/sponsorship-policies).
- `paymasterAddress`:  paymaster address, see [Pimlico addresses docs](https://docs.pimlico.io/paymaster/erc20-paymaster/contract-addresses).
- `paymasterTokenAddress`: paymaster ERC20 token address.
- `amountToApprove`: optional parameter to set the paymaster approve amount

```typescript

export type paymasterOptions = {
  paymasterUrl?: string
  isSponsored?: boolean
  sponsorshipPolicyId?: string
  paymasterAddress: string
  paymasterTokenAddress?: string
  amountToApprove?: bigint
}


```

## How to Test

Added 6 scripts to tests all 4337 features in the playground folder: 

- `usdc-transfer-4337.ts`
- `usdc-transfer-4337-erc20.ts`
- `usdc-transfer-4337-sponsored.ts`
- `usdc-transfer-4337-counterfactual.ts`
- `usdc-transfer-4337-erc20-counterfactual.ts`
- `usdc-transfer-4337-sponsored-counterfactual.ts`

### CODE REVIEW

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 1]
```diff
@@ -520,18 +546,4 @@ export class Safe4337Pack extends RelayKitBasePack<{
       )
     ])
   }
-
-  #userOperationToHexValues(userOperation: UserOperation) {
```
Comment (DaniSomoza) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:524]: I Added this function as an util to be able to use it in the estimation files

File: playground/relay-kit/usdc-transfer-4337.ts [hunk 2]
```diff
@@ -34,19 +35,20 @@ async function main() {
...
+      safeAddress: SAFE_ADDRESS
     }
   })
 
   // Log supported entry points and chain id
   console.log('Supported Entry Points', await safe4337Pack.getSupportedEntryPoints())
   console.log('Chain Id', await safe4337Pack.getChainId())
 
-  // Create transaction batch to transfer 2 USDC
+  // Create transaction batch with 2 USDC transfers
```
Comment (dasanra) [inline playground/relay-kit/usdc-transfer-4337.ts:46]:
```suggestion
  // Create transaction batch with 0.2 USDC transfers
```

File: playground/relay-kit/usdc-transfer-4337-paymaster.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    options: {
+      safeAddress: SAFE_ADDRESS
+    }
+  })
+
+  // Log supported entry points and chain id
+  console.log('Supported Entry Points', await safe4337Pack.getSupportedEntryPoints())
+  console.log('Chain Id', await safe4337Pack.getChainId())
+
+  // Create transaction batch to transfer 2 USDC
```
Comment (dasanra) [inline playground/relay-kit/usdc-transfer-4337-paymaster.ts:58]:
```suggestion
  // Create transaction batch with two 0.1 USDC transfers
```

File: playground/relay-kit/usdc-transfer-4337-counterfactual.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  const fundingSafe = {
+    to: senderAddress,
+    value: ethers.parseEther(nativeTokenAmount)
+  }
+
+  console.log(`sending ${nativeTokenAmount} ETH...`)
+
+  await signer.sendTransaction(fundingSafe)
+
+  // Create transaction batch with 2 USDC transfers
```
Comment (dasanra) [inline playground/relay-kit/usdc-transfer-4337-counterfactual.ts:62]:
```suggestion
  // Create transaction batch with two 0.1 USDC transfers
```

File: playground/relay-kit/usdc-transfer-4337-counterfactual-paymaster.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+  console.log('senderAddress: ', senderAddress)
+
+  // funding the Safe with USDC
+
+  console.log(`sending 15 USDC...`)
+
+  await transfer(signer, usdcTokenAddress, senderAddress, 15_000_000n)
+
+  // Create transaction batch with 2 USDC transfers
```
Comment (dasanra) [inline playground/relay-kit/usdc-transfer-4337-counterfactual-paymaster.ts:67]:
```suggestion
  // Create transaction batch with two 0.1 USDC transfers
```

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 6]
```diff
@@ -220,14 +285,36 @@ export class Safe4337Pack extends RelayKitBasePack<{
...
+        safeOperation.addEstimations(adjustEstimationData)
+      }
+    }
+
+    if (isPaymasterPresent) {
+      if (!this.#PAYMASTER_URL) {
+        throw new Error('No paymaster url provided')
+      }
+
+      const paymasterEstimation = await feeEstimator?.getPaymasterEstimation?.({
```
Comment (dasanra) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:309]: There is a typo here and I suggest a different naming convention
```suggestion
    const usePaymaster = userOperation.paymasterAndData !== '0x'
    const hasSafeDeployment = userOperation.initCode !== '0x'

    // adjustment only needed for deployments without paymaster
    if (!usePaymaster && hasSafeDeployment) {
      const adjustEstimationData = await feeEstimator?.adjustEstimation?.({
        bundlerUrl: this.#BUNDLER_URL,
        entryPoint: this.#ENTRYPOINT_ADDRESS,
        userOperation: safeOperation.toUserOperation()
      })

      if (adjustEstimationData) {
        safeOperation.addEstimations(adjustEstimationData)
      }
    }

    if (usePaymaster) {
      if (!this.#PAYMASTER_URL) {
        throw new Error('No paymaster url provided')
      }

      const paymasterEstimation = await feeEstimator?.getPaymasterEstimation?.({
```

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 7]
```diff
@@ -240,11 +327,35 @@ export class Safe4337Pack extends RelayKitBasePack<{
...
+        to: erc20TokenAddress,
+        data: INTERFACES.encodeFunctionData('approve', [this.#paymasterAddress, amountToApprove]),
+        value: '0',
+        operation: OperationType.Call // Call for approve
+      }
+
+      transactions.push(approveToPaymasterTransaction)
+    }
+
+    const usePaymaster = options?.usePaymaster !== false && !!this.#paymasterAddress
```
Comment (dasanra) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:355]: You already throw an error if `#paymasterAddress` is not defined, why not to just use the option as the flag?

Also you are adding the approve transaction above if they use a `erc20TokenAddress` regardless of the `usePaymaster` option is true or false. 

I think you can improve this code section

Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:355]: > You already throw an error if #paymasterAddress is not defined, why not to just use the option as the flag?

The `options?.usePaymaster` flag allows users to choose to use the paymaster for a transaction or not. The check that throws the exception ensures that this flag isn't used without first initializing a `paymasterAddress`.

> Also you are adding the approve transaction above if they use a `erc20TokenAddress` regardless of the usePaymaster option is true or false.

True! I updated the condition to include the `usePaymaster` flag:

```typescript
if (options?.erc20TokenAddress && usePaymaster) {

```

File: playground/relay-kit/usdc-transfer-4337-counterfactual-paymaster.ts (hunk 5 above)
Comment (DaniSomoza) [reply playground/relay-kit/usdc-transfer-4337-counterfactual-paymaster.ts]: done!

File: playground/relay-kit/usdc-transfer-4337-counterfactual.ts (hunk 4 above)
Comment (DaniSomoza) [reply playground/relay-kit/usdc-transfer-4337-counterfactual.ts]: done!

File: playground/relay-kit/usdc-transfer-4337-paymaster.ts (hunk 3 above)
Comment (DaniSomoza) [reply playground/relay-kit/usdc-transfer-4337-paymaster.ts]: done!

File: playground/relay-kit/usdc-transfer-4337.ts (hunk 2 above)
Comment (DaniSomoza) [reply playground/relay-kit/usdc-transfer-4337.ts]: done!

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts (hunk 6 above)
Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts]: done!

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 8]
```diff
@@ -141,6 +154,45 @@ export class Safe4337Pack extends RelayKitBasePack<{
         throw new Error('Owners and threshold are required to deploy a new Safe')
       }
 
+      const isPaymasterPresent =
+        paymasterOptions && paymasterOptions.paymasterAddress && paymasterOptions.erc20TokenAddress
+
```
Comment (yagopv) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:159]: What do you think about renaming the erc20TokenAddress prop to something more meaningful as paymasterTokenAddress

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 9]
```diff
@@ -220,14 +274,36 @@ export class Safe4337Pack extends RelayKitBasePack<{
...
+        userOperation: safeOperation.toUserOperation()
+      })
+
+      if (adjustEstimationData) {
+        safeOperation.addEstimations(adjustEstimationData)
+      }
+    }
+
+    if (usePaymaster) {
+      if (!this.#PAYMASTER_URL) {
```
Comment (yagopv) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:294]: Is the paymaster url different than the bundler one? Just checking if the prop is needed or the url is going to remain the same across providers

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 10]
```diff
@@ -95,7 +107,8 @@ export class Safe4337Pack extends RelayKitBasePack<{
    * @return {Promise<Safe4337Pack>} The Promise object that will be resolved into an instance of Safe4337Pack.
    */
   static async init(initOptions: Safe4337InitOptions): Promise<Safe4337Pack> {
```
Comment (yagopv) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:109]: The fn is long now. Perhaps we can group it in smaller ones

**Comment by yagopv** [comment]
`@param {EstimateFeeProps{EstimateFeeFn}}` this comment is obsolete. The param is a FeeEstimator now

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts [hunk 11]
```diff
@@ -240,11 +316,35 @@ export class Safe4337Pack extends RelayKitBasePack<{
    * @return {Promise<SafeOperation>} The Promise object will resolve a SafeOperation.
    */
   async createTransaction({
-    transactions
+    transactions,
+    options = {}
```
Comment (yagopv) [inline packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:320]: Review JSdoc for this fn. Missing options prop and the style is different from the `getEstimateFee` method for example

File: packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts (hunk 9 above)
Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/Safe4337Pack.ts:294]: Yes, it seems that now in the docs, they are using the same v2 endpoint for both clients:

```
https://docs.pimlico.io/permissionless/tutorial/tutorial-1#create-the-bundler-and-smart-account-clients-and-fetch-the-gas-price
```

File: packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts [hunk 12]
```diff
@@ -19,6 +26,26 @@ export class PimlicoFeeEstimator implements IFeeEstimator {
...
+  }: EstimateSponsoredFeeFunctionProps): Promise<EstimateSponsoredGasData> {
+    const paymasterClient = new ethers.JsonRpcProvider(paymasterUrl, undefined, {
+      batchMaxCount: 1
+    })
+
+    const params = [userOperationToHexValues(userOperation), entryPoint]
+
+    const gasEstimate = await paymasterClient.send(
+      'pm_sponsorUserOperation',
+      sponsorshipPolicyId ? [...params, { sponsorshipPolicyId }] : params
```
Comment (dasanra) [inline packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts:43]: You add this rule inside the send function when the params array is created 2 lines above. Wouldn't make sense to do this check when creating the element?

`params` is not used later for anything else

File: packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts [hunk 13]
```diff
@@ -19,6 +26,26 @@ export class PimlicoFeeEstimator implements IFeeEstimator {
...
+    sponsorshipPolicyId
+  }: EstimateSponsoredFeeFunctionProps): Promise<EstimateSponsoredGasData> {
+    const paymasterClient = new ethers.JsonRpcProvider(paymasterUrl, undefined, {
+      batchMaxCount: 1
+    })
+
+    const params = [userOperationToHexValues(userOperation), entryPoint]
+
+    const gasEstimate = await paymasterClient.send(
+      'pm_sponsorUserOperation',
```
Comment (yagopv) [inline packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts:42]: Perhaps we can add this and the one in the PimlicoEstimator in the constants file same as other RPC calls

**Review by yagopv — APPROVED** [review]

**Comment by yagopv** [comment]
Great work!

File: packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts (hunk 12 above)
Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts:43]: we can do something like this instead:

```typescript

    const params = sponsorshipPolicyId
      ? [userOperationToHexValues(userOperation), entryPoint, { sponsorshipPolicyId }]
      : [userOperationToHexValues(userOperation), entryPoint]

    const gasEstimate = await paymasterClient.send('pm_sponsorUserOperation', params)
    
```

Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts]: done!

File: packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts (hunk 13 above)
Comment (DaniSomoza) [reply packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts]: done!

File: packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts (hunk 12 above)
Comment (dasanra) [reply packages/relay-kit/src/packs/safe-4337/estimators/PimlicoFeeEstimator.ts]: Duplicated two params, but for me it improved the legibility

File: packages/relay-kit/src/packs/safe-4337/constants.ts [hunk 14]
```diff
@@ -30,5 +31,6 @@ export const RPC_4337_CALLS = {
   GET_USER_OPERATION_BY_HASH: 'eth_getUserOperationByHash',
   GET_USER_OPERATION_RECEIPT: 'eth_getUserOperationReceipt',
   SUPPORTED_ENTRY_POINTS: 'eth_supportedEntryPoints',
-  CHAIN_ID: 'eth_chainId'
+  CHAIN_ID: 'eth_chainId',
+  SPONSOR_USER_OPERATION: 'pm_sponsorUserOperation'
```
Comment (dasanra) [inline packages/relay-kit/src/packs/safe-4337/constants.ts:35]: This is Pimlico specific but set in global constants. Not part of the 4337 specification.

**Review by dasanra — APPROVED** [review]

**Pull request merged**